grpclog/loggerv2_test.go

```diff
@@ -61,3 +66,220 @@ func checkLogForSeverity(s int, b []byte) error {
+	})
+
+	t.Run("no discard", func(t *testing.T) {
+		buffers := []*bytes.Buffer{new(bytes.Buffer), new(bytes.Buffer), new(bytes.Buffer)}
+		logger := NewLoggerV2(buffers[infoLog], buffers[warningLog], buffers[errorLog])
+		loggerTp := logger.(*loggerT)
+
+		loggerTp.printFunc(errorLog, logFuncOk)
+
+		infoB, err := buffers[infoLog].ReadBytes('\n')
+		if err != nil {
+			t.Fatalf("level %v: %v", infoLog, err)
+		}
+		checkLogContainsFuncStr(infoB)
+
+		warnB, err := buffers[warningLog].ReadBytes('\n')
+		if err != nil {
+			t.Fatalf("level %v: %v", warningLog, err)
+		}
+		checkLogContainsFuncStr(warnB)
+
+		errB, err := buffers[errorLog].ReadBytes('\n')
+		if err != nil {
+			t.Fatalf("level %v: %v", errorLog, err)
+		}
+		checkLogContainsFuncStr(errB)
+	})
+
+	t.Run("all discard", func(t *testing.T) {
+		logger := NewLoggerV2(io.Discard, io.Discard, io.Discard)
+		loggerTp := logger.(*loggerT)
+
+		loggerTp.printFunc(errorLog, logFuncErr)
+	})
+
+	t.Run("all combinations", func(t *testing.T) {
+		const (
+			print int = iota
+			printf
+			println
+		)
+		type testDiscard struct {
+			discardInf  bool
+			discardWarn bool
+			discardErr  bool
+		}
+		cfgs := []loggerV2Config{
+			{jsonFormat: false},
+			{jsonFormat: true},
+		}
+		discardTests := []testDiscard{
+			{discardInf: true, discardWarn: true, discardErr: true},
+			{discardInf: true, discardWarn: true, discardErr: false},
+			{discardInf: true, discardWarn: false, discardErr: true},
+			{discardInf: false, discardWarn: true, discardErr: true},
+			{discardInf: true, discardWarn: false, discardErr: false},
+			{discardInf: false, discardWarn: false, discardErr: true},
+			{discardInf: false, discardWarn: false, discardErr: false},
+		}
+		for _, cfg := range cfgs {
+			for printType := print; printType <= println; printType++ {
+				for _, discardTest := range discardTests {
+					buffers := []*bytes.Buffer{new(bytes.Buffer), new(bytes.Buffer), new(bytes.Buffer)}
+					writers := []io.Writer{buffers[infoLog], buffers[warningLog], buffers[errorLog]}
+					if discardTest.discardInf {
+						writers[infoLog] = io.Discard
+					}
+					if discardTest.discardWarn {
+						writers[warningLog] = io.Discard
+					}
+					if discardTest.discardErr {
+						writers[errorLog] = io.Discard
+					}
+					logger := newLoggerV2WithConfig(writers[infoLog], writers[warningLog], writers[errorLog], cfg)
+
+					msgInf := "someinfo"
+					msgWarn := "somewarn"
+					msgErr := "someerr"
+					if printType == print {
+						logger.Info(msgInf)
+						logger.Warning(msgWarn)
+						logger.Error(msgErr)
+					} else if printType == printf {
+						logger.Infof("%v", msgInf)
+						logger.Warningf("%v", msgWarn)
+						logger.Errorf("%v", msgErr)
+					} else if printType == println {
+						logger.Infoln("%v", msgInf)
+						logger.Warningln("%v", msgWarn)
+						logger.Errorln("%v", msgErr)
+					}
+
+					checkWrite := func(buf *bytes.Buffer, discard bool, logType string, msg string, isJson bool) {
```

Comment:
does checkWrite needs to be defined within the loop? Probably good to define all functions at the top